[2단계 - 장기] 레오(김준영) 미션 제출합니다.

kjyyjk opened the pull request:
## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
1단계보다는 한 층 나아졌다고 생각합니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

안녕하세요 이프~
2단계 리뷰도 잘부탁드립니다.😊

이번 2단계에서는 궁성 구현과 DB 접목을 주로 개발했습니다.

### 궁성
궁성을 구현함에 있어 지난 1단계에서 말씀드렸던 전략 패턴을 사용해볼까 고민해보았습니다.
하지만 오로지 궁성 내부와 궁성 외부로 기물의 움직임이 나뉘는데 과연 이 시점에 전략 패턴이 필요할까 의문이 들었습니다.
리팩토링 시간이 부족한 것도 있고 현재 단계에서는 불필요하다고 판단했기에 가장 간단한 방법으로 구현했습니다.

기존에는 각 기물이 ```POSSIBLE_MOVEMENTS```를 가지고 있었는데요. 
궁성을 구현하면서는 궁성 움직임에 변화가 생기는 기물들에게만 ```POSSIBLE_CASTLE_MOVEMENTS```필드를 추가해주었습니다. 
이후 추상 메서드인 ```Piece#getPossibleMovements```를 재정의하며 자신의 위치가 궁성이라면 
```POSSIBLE_CASTLE_MOVEMENTS```를 반환하게끔 구현했습니다.

### DB
DB도 가장 빠르게 요구사항을 만족할 수 있게끔 설계하고 구현했습니다.
`game 테이블` : `id(pk)`, `team(현재 턴의 차례)`
`board_piece 테이블` : `id(pk)`, `game_id(fk)`, `column_value`, `row_value`, `piece_type`, `team`

현재 단계에서는 동시에 하나의 게임만을 진행할 수 있게 하고자,
Application에서 각 DAO 객체의 인자로 `game_id` 1을 넘기게끔 하드 코딩했습니다.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **Documentation**
  - 업데이트된 게임 설명서에 로컬 MySQL 데이터베이스 설정과 함께 게임 준비, 진행, 기물 이동 및 종료 규칙에 대한 상세 안내가 추가되었습니다.

- **New Features**
  - 게임 진행 시 기존 데이터가 자동으로 불러와지며, 오류 처리를 강화하여 더욱 안정적인 플레이 환경을 제공합니다.
  - 각 기물의 이동 옵션 및 점수 체계가 개선되어 전략적 게임 진행이 강화되었으며, 시작 및 결과 메시지가 보다 명확하게 표시됩니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kjyyjk committed:
docs: 궁성과 점수 계산에 대한 기능 목록 추가

kjyyjk committed:
test(red): 초나라 궁이 궁성을 벗어날 시 예외를 발생

kjyyjk committed:
feat(green): 초나라 궁이 궁성을 벗어날 시 예외를 발생

kjyyjk committed:
test(red): 한나라 궁이 궁성을 벗어날 시 예외를 발생

kjyyjk committed:
feat(green): 한나라 궁이 궁성을 벗어날 시 예외를 발생

kjyyjk committed:
refactor: 한나라 궁이 궁성을 벗어날 시 예외를 발생

kjyyjk committed:
test(red): 초나라 궁이 선없는 대각 이동 시 예외를 발생

kjyyjk committed:
feat(green): 초나라 궁이 선없는 대각 이동 시 예외를 발생

kjyyjk committed:
refactor: 초나라 궁이 선없는 대각 이동 시 예외를 발생

kjyyjk committed:
test(red): 한나라 궁이 선없는 대각 이동 시 예외를 발생

kjyyjk committed:
feat(green): 한나라 궁이 선없는 대각 이동 시 예외를 발생

kjyyjk committed:
test(red): 초나라 궁이 선따라 대각 이동 가능

kjyyjk committed:
feat(green): 초나라 궁이 선따라 대각 이동 가능

kjyyjk committed:
feat: 한나라 궁이 선따라 대각 이동 가능

kjyyjk committed:
test: 궁이 두칸 이동 시 예외를 발생

kjyyjk committed:
feat: 사가 궁과 같은 움직임 조건을 가지도록 구현

kjyyjk committed:
refactor: 공통 메서드의 테스트는 Piece 테스트로 이동

kjyyjk committed:
feat: 졸병은 뒤로 이동 불가

kjyyjk committed:
test: 졸병이 궁성 내에서 대각으로 뒤를 향해 움직이면 예외 발생

kjyyjk committed:
test(red): 졸병이 궁성 내에서 앞을 향해 대각으로 움직이는 기능

kjyyjk committed:
feat(green): 졸병이 궁성 내에서 앞을 향해 대각으로 움직이는 기능 구현

kjyyjk committed:
refactor: 선 없는 궁성 대각 움직임 예외 발생을 상위에서 공통 처리

kjyyjk committed:
feat: 졸병이 궁성 밖에서 대각으로 움직이면 예외 발생

kjyyjk committed:
feat: 차가 궁성 내에서 상/하/좌/우/대각으로 최대 2칸 움직이는 기능 구현

kjyyjk committed:
feat: 포가 궁성 내에서 상/하/좌/우/대각으로 2칸 움직이는 기능 구현

kjyyjk committed:
feat: 게임 종료 시 각 진영의 점수 계산 및 출력

kjyyjk committed:
chore: mysql-connector 의존성 추가

kjyyjk committed:
docs: 데이터베이스 관련 기능 목록 추가

kjyyjk committed:
test(red): 기물 종류에 따른 PieceType 반환

kjyyjk committed:
feat(green): 기물 종류에 따른 PieceType 반환

kjyyjk committed:
test(red): PieceType에 따른 기물 생성

kjyyjk committed:
feat(green): PieceType에 따른 기물 생성

kjyyjk committed:
feat: DB에 게임 정보를 저장 및 관리

- 이미 진행중인 게임이 있으면 진행 정보를 불러와 보드를 구성한다.

- 진행중인 게임이 존재하지 않으면 게임과 보드를 초기화하여 db에 저장한다.

- 매 턴 시작마다 DB에서 게임의 턴(진영) 정보를 불러온다.

- 매 턴 종료마다 DB에 게임의 턴(진영) 정보를 업데이트 한다.

- 매 기물 이동마다 DB에 기물 정보를 업데이트 한다.

- 게임 종료 시 게임 진행 정보를 전부 삭제한다.

create table game (
	id int NOT NULL,
    turn ENUM('GREEN', 'RED') NOT NULL,
    CONSTRAINT game_pk PRIMARY KEY(id)
);

create table board_piece (
	id int NOT NULL AUTO_INCREMENT,
    game_id int NOT NULL,
    column_value int NOT NULL,
    row_value int NOT NULL,
    piece_type ENUM('SOLDIER', 'GUARD', 'ELEPHANT', 'HORSE', 'CANON', 'CHARIOT', 'GENERAL') NOT NULL,
    team ENUM('GREEN', 'RED') NOT NULL,
	CONSTRAINT board_piece_pk PRIMARY KEY(id),
    CONSTRAINT game_board_piecefk foreign key (game_id) references game (id)
);

kjyyjk committed:
docs: README에 DB 세팅 방법 작성

kjyyjk committed:
refactor: InputView 내부에서 InputParser를 사용

kjyyjk committed:
fix: test 케이스 오타 수정

kjyyjk committed:
refactor: try-with-resources 사용

- 기존에는 Connection, PreparedStatement, ResultSet을 사용후에
close를 해주지 않고 있느나 모두 사용 후 반납 해주어야하는 자원들이다.

- try-with-resources를 사용하면 명시적으로 close를 하지 않아도 묵시적으로 close된다.

kjyyjk committed:
refactor: 게임 ID를 외부에서 주입

- 동시에 하나의 게임만 가능하게 하도록 기존에는 dao 객체의 쿼리에
game_id를 모두 1로 고정해놨었다. 하지만 추후 여러 게임을 관리 가능한 구조를
가져가려면 외부에서 gameId를 인자로 주입해주는게 맞다고 판단했다.
따라서 현재는 하나의 게임만 관리하돼, 외부에서 1이라는 하드코딩 된 값을 주입해주도록
수정했다.

kjyyjk commented on src/main/java/janggi/PieceType.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+public enum PieceType {
+    GENERAL(Piece::isGeneral, General::new),
+    GUARD(Piece::isGuard, Guard::new),
+    HORSE(Piece::isHorse, Horse::new),
+    ELEPHANT(Piece::isElephant, Elephant::new),
+    CANON(Piece::isCanon, Canon::new),
+    CHARIOT(Piece::isChariot, Chariot::new),
+    SOLDIER(Piece::isSoldier, Soldier::new);
+
+    private final Predicate<Piece> typeCondition;
+    private final Function<Team, Piece> pieceCreator;
+
+    PieceType(Predicate<Piece> typeCondition, Function<Team, Piece> pieceCreator) {
+        this.typeCondition = typeCondition;
+        this.pieceCreator = pieceCreator;
+    }
+
+    public static PieceType from(Piece piece) {
+        for (PieceType pieceType : values()) {
+            Predicate<Piece> condition = pieceType.typeCondition;
+            if (condition.test(piece)) {
+                return pieceType;
+            }
+        }
+        throw new IllegalArgumentException("[ERROR] 존재하지 않는 타입의 기물입니다.");
+    }
+
+    public Piece toPiece(Team team) {
+        return pieceCreator.apply(team);
+    }
+}
```
`board_piece` 테이블에 기물을 저장할 때 해당 기물의 종류가 무엇인지 구분할 필요가 있었습니다.
따라서 처음에는 `Piece` 객체에 각 타입을 나타내는 필드를 추가할까 고민했지만,
이전에 사용하지 않다가 상태를 추가하는 것은 도메인 객체와 DB의 결합도를 증가시키는 것이라고 판단했습니다.

도메인 객체가 DB의 설계에 종속되지 않길 원해서 이와 같은 `PieceType` Enum 클래스를 활용했습니다.
`PieceType#from`으로 각 Piece 하위 클래스를 PieceType으로 변환하거나
`PieceType#toPiece`로 DB에서 읽어온 PieceType으로부터 해당 타입의 구현체를 생성할 수 있게끔 구현했습니다. 

이러한 제 의도와 해결방법에 대해 이프는 어떻게 생각하시는지 궁금합니다!

sinb57 replied:
PieceType Enum 클래스 사용은 잘 하셨다고 생각해요. PieceType의 typeCondition에 대해서는 부정적으로 생각해요.

기물 타입이 추가될 때마다 `Piece` 추상 클래스에 `isSoldier` 와 같은 메서드가 계속 추가되고, 하위의 구현 클래스들이 모두 이를 override 해야 할텐데요. OCP를 위배한 것은 아닌지 고민해보시면 좋겠어요.

kjyyjk replied:
우선 OCP는 무엇을 의미하는 지 다시 알아보았고 제가 생각한 OCP는 다음과 같습니다.
```
추상화와 긴밀하게 연결되어있는 개념이다.
적절한 추상화를 통해 새로운 하위 클래스를 추가해도(확장o) 기존의 코드를 수정하지 않도록(변경x) 하는 것이다
```

이러한 OCP에 비교했을 때 제 기존 코드는 확장과 변경에 모두 열려있는 구조였습니다.
`기물이 추가될 때마다(확장o) Piece에 is하위기물 이라는 메서드가 추가된다.(변경o)`

그렇다면 왜 OCP를 지켜야할까 생각해보았습니다.
OCP를 지키지 않아 현재 구조를 계속 가져가면 기물의 종류를 추가할 때마다 Piece에 메서드가 추가되어야 합니다. 이 경우 기존 코드를 수정한다는 점에서 사이드 이펙트가 발생할 수 있습니다. 또한 Piec 클래스가 점점 비대해져 유지보수가 힘들어질 것 같습니다.

그럼 어떻게 해결할 수 있을까를 고민해보았을 때
생각나는 방법은 제가 지양하고자 했던 PieceType을 Piece의 상태에 추가하는 것 뿐이었는데요.

그럼 왜 해당 방식을 지양해야할지 다시 생각해보았고, 
현재 코드의 OCP 위배 문제는 DB 접목 이전부터 발생하고 있었던 것이라는 것을 깨달았습니다. 
그렇기 때문에 `PieceType`을 각 하위 클래스에 추가한다고 해서 도메인을 DB에 종속시키는 것이 아니라는 결론을
내렸습니다.

따라서 각 하위 Piece에 PieceType을 추가하는 방향으로 리팩토링 했습니다!
aee03d6ca303c0a381beab59c988d7287566b3f8

kjyyjk replied:
추가로 궁금한 것이 있습니다!
리팩토링을 했지만 결국 기물을 새로 추가함에 따라 `PieceType`와 `PieceCreator`에 코드를 추가해줘야 하는데요.
이 부분은 OCP에서 말하는 `변경`이 아닌걸까요?

우선 저는 기물 종류 확장에 따른 타입 추가와 생성 로직 추가는 `확장`의 범위 내라고 생각했습니다.

sinb57 replied:
기존의 코드가 문제 되었던 부분은, 새로운 타입의 하위 구현 클래스가 추가될 때 `isSoldier`와 같은 메서드가 모든 클래스에게 구현되어야 한다는 점이었어요. 클래스 하나 추가했으나 모든 클래스에 메서드를 추가해야 하는 상황이었던 것이죠.

그에 반해 지금 구조는, 타입 추가가 훨씬 간결해졌어요. `PieceType`와 `PieceCreator`에 변경이 이뤄나는 것을 걱정하셨지만, 적어도 추가하고자 하는 타입 외 모든 클래스에 변경이 이뤄나진 않잖아요. `PieceType`와 `PieceCreator`에 변경이 이뤄나는 것은 자연스럽다고 보고 있고, 변경의 여파를 어떻게 제어할 수 있는지 판단하여 선택하시면 돼요. 설계와 구현은 매 순간이 트레이드 오프예요

kjyyjk commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -21,11 +22,22 @@ public boolean isDifferentTeam(Team team) {
     }
 
     public void validateMovable(Board board, Position start, Position goal) {
-        Path path = calculatePath(start, goal);
+        PossibleMovements possibleMovements = getPossibleMovements(board, start);
+        Path path = possibleMovements.calculatePath(start, goal);
+        validateInvalidCastleDiagonalMove(board, start, goal, path);
         validatePath(board, path);
         validatePieceOnGoal(board, goal);
     }
```
현재는 `Board`에서 이 `Piece#validateMovable`을 호출해줌으로써 기물이 이동 가능한지를 검증하고 있습니다.
그리고 해당 메서드가 너무 많은 작업을 하고 있다고 판단해서 리팩토링 대상으로 고려중입니다.
(갈 수 있는지 검증/경로 계산/보드판에 선이 존재하는지 검증/경로 검증/도착 지점 검증...)

왜 하나의 메서드가 여러 작업을 하고 있으면 안될까라는 물음에 대한 답을 생각해보았을 때,
```
1. 여러 작업이 하나의 메서드 내부에 긴밀하게 결합되어있기 때문에 어떤 하나의 작업에 변경 발생 시,
 다른 작업에게도 변경이 전파되기 쉽다.
2. 가독성이 좋지 않아 한눈에 코드를 이해하기가 힘들다.
```
이 두 가지를 근거로 삼았습니다.

만약 이를 리팩토링한다면 
```
사용자가 출발 지점과 목적 지점을 입력
-> Board가 출발 지점의 기물과 목적 지점의 기물이 같은 팀인지 검증
-> Board가 Piece에게 출발 지점과 목적 지점을 인자로 건내며 Path를 반환할 것을 요구
-> Piece는 자신의 이동 전략으로 출발 지점부터 목적 지점까지 갈 수 있는 지를 검증하고 가능하다면 Path를 반환
-> Board는 Path가 보드판 선에 따라 유효한 지를 검증
-> Board가 Path 상에 위치한 Piece들을 해당 Piece에게 인자로 건내며 경로를 검증할 것을 요구
-> Piece는 경로 상 다른 Piece 정보들을 받아 자신의 움직임 로직에 따라 검증
```
이런 식으로 Board와 Piece간에 메세지를 여러 번 주고 받을 수 있게 함으로써 하나의 메서드 크기를 작게 만들 계획입니다.

아마 리팩토링을 완료하기 이전에 이프가 리뷰를 달아주실 것 같아 한번 확인?받아보고 싶습니다!

kjyyjk replied:
또한 학습적인 측면으로 여쭤보고싶은 것이 있는데요,
```
1. 여러 작업이 하나의 메서드 내부에 긴밀하게 결합되어있기 때문에 어떤 하나의 작업에 변경 발생 시,
 다른 작업에게도 변경이 전파되기 쉽다.
2. 가독성이 좋지 않아 한눈에 코드를 이해하기가 힘들다.
```
제가 말씀드렸던 근거들이 너무 교과서적인? 그러한 것들이 아닐까하는 생각이 듭니다.

이러한 상황에서 이프는
`왜?`라는 질문을 더 던져가며 자신의 가치관이 담긴 근거들을 찾아가시는 편인가요?
아니면 리팩토링에 대한 근거로 이 정도 고려해봤으면 충분하다고 생각하시나요?

깊이 있는 학습을 해보고 싶은 욕심은 있는데 아직은 방법을 잘 모르겠어서 어려운 것 같아요.

sinb57 replied:
`사이드 이펙트를 유추하기 어렵다`와 `가독성 저하`를 근거로 삼으셨군요!! 이 문제를 해결하기 위해선 어떻게 개선할 수 있을까요?

---

레벨1이 마무리된 지금, 레벨1을 돌이켜보고 레벨2를 준비하는 마음으로 한번 고민해보길 바라며 질문을 던져보아요.

우리가 클래스를 분리하는 이유는 무엇일까요? 왜 메서드를 분리할까요? 객체지향을 왜 할까요?? 이러한 부분에 대한 충분한 철학적 고민이 이뤄지는 것이 좋다고 생각해요. 항상 객체지향이 좋은 것은 아니에요. 클래스가 많아지고 코드가 방대해질 수록 오히려 구조를 파악하기 어려울 수도 있어요. 그럼에도 오늘날 객체지향을 지향하는 이유는 무엇일까요?

객체지향 생활체조 원칙은 무엇을 의도한 규칙들일까요? 현업에서는 절대 적용하지 못할 텐데, 무엇을 깨우치길 바라고서 만든 규칙일까요? 레오는 10줄을 넘기지 말라는 규칙에서 무엇을 배웠나요? 2 depth를 넘기지 말라는 규칙에서 무엇을 배웠나요? 스스로 규칙을 잘 따르고 있었는지, 혹은 제약 조건에만 얽매어서 억지스러운 코드를 짜내지는 않았는지 한번 돌이켜 보셨으면 좋겠어요.

이러한 고민들이 모두 해결되어야 비로소 객체지향 관점의 리팩터링에 눈을 뜰 수 있어요. 그 전까지는 그저 private 메서드만 많아지는 데이터 중심으로의 절차지향적인 코드만 탄생할 가능성이 높아요.

레오가 생각하는 `객체지향`은 무엇이고 `객체`는 무엇이며, `책임`과 `변경`의 관계는 무엇인가요?

sinb57 commented on src/main/java/janggi/dao/BoardPieceDao.java:
```diff
@@ -0,0 +1,127 @@
+        preparedStatement.setString(4, pieceType.name());
+        preparedStatement.setString(5, piece.getTeam().name());
+        preparedStatement.executeUpdate();
+    }
```
하나의 `preparedStatement`를 계속 재사용하는 구조가 좋은지 잘 모르겠어요. 이런 방식은 자칫 이전의 executeUpdate 조건이 다음 executeUpdate 조건에 영향을 줄 수 있어요. 되도록 독립적으로 로직을 구성해보세요.

kjyyjk replied:
이프가 말씀해주신 구조로 변경한다면 쿼리를 실행할 때마다 매번 동일한 PreparedStatement 객체를 생성하고 닫아주는 자원 관리가 필요한데 이게 과연 효율적인가 의구심이 듭니다..!

그럼에도 이프가 말씀주신 `조건에 영향을 준다`라는 부분에는 동의해서 한번 사용이 끝난 이후에는
`preparedStatement#clearParameters`를 호출해 조건들을 초기화 해주었는데요!

이에 대해서는 어떻게 생각하시는지 궁금합니다~~
b9436de5a63af2581fc436e6b78b1d33889cca8e

kjyyjk replied:
또한 보드의 피스 현황을 저장하는 것은 하나의 트랜잭션 내에서 이루어지는 것이 맞다고 생각해서 수동 커밋을 활용하도록 수정했습니다.

sinb57 replied:
`preparedStatement#clearParameters`을 통해 조건을 초기화하고 계시는데, preparedStatement를 매번 새로 생성하길 권장드려요.

자원 재사용은, 코드 변경 시 어떠한 영향을 줄 지 사이드 이펙트를 감히 고려할 수 없기 때문에 지양해야 해요. 이러한 요소는 변경을 닫히게 만들기 때문에, 기능 확장이나 수정을 어렵게 만들어요. 이는 어플리케이션의 성장에 큰 걸림돌이 돼요. 우리가 흔히 접하는 `불변 객체`가 이러한 이유 때문에 대두되어 나온 개념이에요.

그에 반해 자원을 재생성하면서 발생하는 생성 비용은, 자원 재사용으로 인한 사이드 이펙트에 비해 관리 비용보다 오히려 저렵해요. 자원 해제만 잘 해주면 GC에 의해 수집되기도 하며, 필요 시 서버 스펙업으로 해결할 수 있어요.

kjyyjk replied:
이번 리뷰를 통해 `상태를 가지는 자원을 재사용하는 것은 지양하는 것이 좋다`라는 점을 새로 알게 되었습니다.
수정했습니다!
df8c881f29d9094dc9a025adcf4bdd74217c223c

sinb57 commented on src/main/java/janggi/board/position/StartAndGoalPosition.java:
```diff
@@ -0,0 +1,19 @@
+package janggi.board.position;
+
+public class StartAndGoalPosition {
```
view에서 사용되는 DTO로 보이는데요. board.position 패키지가 옳은지 고민해보시면 좋겠어요.

kjyyjk replied:
view 패키지로 이동하는 것이 적절할 것 같아 수정했습니다!
3f297aedb76a07534dacfaed6ac41e280c61142b

sinb57 commented on src/main/java/janggi/dao/BoardPieceDao.java:
```diff
@@ -0,0 +1,127 @@
+        }
+    }
+
+    public void delete(int gameId, Position goalPosition) {
```
BoardPieceDao 입장에서 파라미터명이 `goalPosition`이어야 할까요? 객체 스스로의 관점이 아니라, 해당 메서드를 사용하는 외부의 시선으로 파라미터명이 지어졌네요.

kjyyjk replied:
인텔리제이의 자동 메서드 추가 기능을 사용하는 과정에서 놓친 부분이네요..!
신경쓰지 못해 죄송합니다ㅜ

sinb57 commented on src/main/java/janggi/dao/BoardPieceDao.java:
```diff
@@ -0,0 +1,127 @@
+            preparedStatement.setInt(2, goalRow.getValue());
+            preparedStatement.setInt(3, gameId);
+            preparedStatement.setInt(4, startColumn.getValue());
+            preparedStatement.setInt(5, startRow.getValue());
```
```suggestion
            preparedStatement.setInt(1, goalPosition.getColumn().getValue());
            preparedStatement.setInt(2, n.getRow().getValue());
            preparedStatement.setInt(3, gameId);
            preparedStatement.setInt(4, startPosition.getColumn().getValue());
            preparedStatement.setInt(5, startPosition.getRow().getValue());
```
변수를 따로 둘 필요는 없어보여요. 오히려 행이 많아져서 읽기 어려웠어요.

sinb57 commented on README.md:
```diff
@@ -2,38 +2,87 @@
-- 기물 배치
-  - [x] 한나라 플레이어가 기물을 배치한다.
-  - [x] 초나라 플레이어가 기물을 배치한다.
+2. 생성한 스키마 내부에서 다음 쿼리들을 순차적으로 실행해 테이블을 생성한다.
```
직접 DB를 세팅하도록 가이드 하지 말고, docker-compose를 통해 세팅을 지원해주세요.

kjyyjk replied:
docker-compose는 처음 사용해보는데,
누가 애플리케이션을 실행하든 제 컴퓨터와 동일한 환경을 쉽게 제공해 줄 수 있겠네요!

수정했습니다.
f1090ebef3abfdeaab36cea4fb0315fc7eb39d15

sinb57 commented on src/main/java/janggi/PieceType.java:
```diff
@@ -0,0 +1,44 @@
+            Predicate<Piece> condition = pieceType.typeCondition;
+            if (condition.test(piece)) {
+                return pieceType;
+            }
```
```suggestion
            if (pieceType.typeCondition.test(piece)) {
                return pieceType;
            }
```
변수를 분리할 필요가 없어보이네요

kjyyjk replied:
한 줄에 점을 하나만 찍는다가 이런 상황을 의미하는 것이 아님을 알고 있음에도 불구하고,
가독성을 해쳐가면서까지 잘못 적용하고 있었던 것 같아요.

수정 완료했고, 스스로 성찰하겠습니다..!

sinb57 commented on src/main/java/janggi/moving/Path.java:
```diff
@@ -20,8 +21,35 @@ public List<Position> getIntermediatePath() {
+        return false;
+    }
+
+    public boolean firstAndLastIn(List<Position> positions) {
```
사용하지 않는 메서드네요

sinb57 commented on src/main/java/janggi/dao/BoardPieceDao.java:
```diff
@@ -0,0 +1,127 @@
+import java.util.Map.Entry;
+
+public class BoardPieceDao {
+    private static final String url = System.getenv("DB_URL");
```
여러 방법 중 `System.getenv` 방식으로 필드를 주입한 궁금해요

kjyyjk replied:
우선 민감 정보를 깃에 올리는 것은 아니라고 생각해서 시스템 환경 변수를 생각했습니다.
그리고 자바 코드에서 시스템 환경 변수를 불러오는 방법을 찾아보았을 때 `System.getenv`가 있어서 사용했습니다.

결국 다른 여러 방법들과 장단점을 비교하기보다는 가장 먼저 생각난 방법을 사용했다고 할 수 있습니다.

kjyyjk replied:
하지만 docker-compose를 활용하도록 리팩터링하는 과정에서 해당 코드는 삭제되었습니다!

sinb57 commented on src/main/java/janggi/PieceType.java:
```diff
@@ -0,0 +1,44 @@
+            Predicate<Piece> condition = pieceType.typeCondition;
+            if (condition.test(piece)) {
+                return pieceType;
+            }
```
stream()을 활용해서 2 depth를 해결해보세요

kjyyjk replied:
위 OCP 위배 문제로 리팩토링을 하다 보니 해당 코드를 지우게 됐습니다.

만약 수정했다면
```
return Arrays.stream(values())
    .filter(type -> type.typeCondition.test(piece))
    .findFirst()
    .orElseThrow(...)
```
와 같이 Stream을 사용해 depth를 해결할 것입니다.

sinb57 reviewed (changes requested):
안녕하세요 레오~ 2단계 진행하느라 고생 많으셨어요!
리팩터링에 대한 고민이 있으셔서, 깊이 리뷰를 남기진 않았어요.
코멘트 확인 부탁드려요~

kjyyjk committed:
refactor: Piece에 PieceType 추가

- 기존 구조는 기물 종류가 추가되면 Piece 클래스에 `is하위기물` 형식의 메서드가 추가되어야한다.
이는 OCP를 위배한다고 판단했다.
따라서 Piece 객체가 PieceType을 가지고 있도록 리팩토링했다.
이를 통해 새로운 기물이 추가되어도 Piece에 메서드를 추가할 필요가 없어졌다.

- 특정 Piece 구현체의 생성자에 PieceType을 인자로 넘기다보면 잘못된 PieceType을 넘길
여지가 생긴다고 판단했다. 따라서 PieceType에 따라 Piece를 생성하여 반환하는 PieceCreator를 사용한다.

kjyyjk committed:
refactor: 보드판 저장 로직 수정

- PreparedStatement를 재사용하는만큼 명시적으로 파라미터를 초기화해주었다.

- 또한 보드판 저장 로직은 하나의 트랜잭션 내에서 이루어져야하므로
수동 커밋을 하도록 수정했다. 5000건 insert문을 기준으로 약 15초의
속도 개선이 일어났다.(16ms -> 1ms)

kjyyjk committed:
refactor: StartAndGoalPosition 패키지 이동

- 해당 객체는 View에서 사용하는 Dto의 일종이므로 기존 board.position 보다는
view 패키지에 위치하는 것이 적절해보여 패키지를 이동했다.

kjyyjk committed:
refactor: 미사용 메서드 제거

kjyyjk committed:
refactor: docker-compose를 활용해 DB 세팅

sinb57 reviewed (changes requested):
안녕하세요 레오~ 리뷰 반영하시느라 고생 많으셨어요!
코멘트 확인해주시면 다음 요청 때 머지할게요~!!

kjyyjk committed:
refactor: PreparedStatement 재사용 로직 제거

- 기존에는 자원 관리비용을 아끼기 위해 PreparedStatement를 재사용했으나,
상태를 가지는 객체를 재사용하는 것은 추후 사이드 이펙트가 발생할 수 있다.
그리고 해당 사이드 이펙트에 비해 객체 관리 비용이 더 저렴하다고 할 수 있어
재사용하지 않게끔 수정한다.

sinb57 reviewed (approved):
안녕하세요 레오! 주말에도 장기 미션 진행하느라 고생 많으셨어요!!
이번 장기 미션과 더불어 지난 2개월의 레벨1에서 많은 것을 학습하셨길 바라면서, 장기 미션은 이만 마루리 짓도록 할게요! 혹여나 리뷰에 대한 추가 질문이나 우테코 생활에 대한 고민이 있다면 언제든 편하게 DM 연락주세요~
